src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,122 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final double GROW_FACTOR = 1.5;
+    private static final int NOT_FOUND = -1;
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elements;
+    private int size;
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            elements = (T[]) new Object[initialCapacity];
+        } else {
+            throw new IllegalArgumentException("Illegal initial capacity: " + initialCapacity);
+        }
+    }
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfNeed();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index " + index);
+        }
 
+        if (index == size) {
+            add(value);
+        } else {
+            growIfNeed();
+            System.arraycopy(elements, index, elements, index + 1, size++ - index);
+            elements[index] = value;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexValid(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexValid(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexValid(index);
+
+        T oldValue = elements[index];
+        if (index == size - 1) {
+            elements[index] = null;
+        }
+
```

Comment:
```suggestion
```
arraycopy will handle this, no need to do it manually

Reply:
I forgot to add return statement. It's need to avoid redundant call of arraycopy.